halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Plugin.java

```diff
@@ -72,4 +85,47 @@ public InputStream downloadManifest() {
               .build());
     }
   }
+
+  public static Map<String, Object> normalizeOptions(Map<String, Object> opts) {
+    Map<String, Object> result = new HashMap<>();
+    for (Map.Entry<String, Object> entry : opts.entrySet()) {
+      String key = entry.getKey();
+      result.putAll(parseOption(key, entry.getValue()));
+    }
+
+    return result;
+  }
+
+  public static Map<String, Object> merge(Map original, Map newMap) {
+    for (Object key : newMap.keySet()) {
+      if (newMap.get(key) instanceof Map && original.get(key) instanceof Map) {
+        Map originalChild = (Map) original.get(key);
+        Map newChild = (Map) newMap.get(key);
+        original.put(key, Plugin.merge(originalChild, newChild));
+      } else if (newMap.get(key) instanceof List && original.get(key) instanceof List) {
+        List originalChild = (List) original.get(key);
+        List newChild = (List) newMap.get(key);
+        for (Object each : newChild) {
+          if (!originalChild.contains(each)) {
+            originalChild.add(each);
+          }
+        }
+      } else {
+        original.put(key, newMap.get(key));
+      }
+    }
+    return (Map<String, Object>) original;
+  }
+
+  public static Map<String, Object> parseOption(String key, Object value) {
+    Map<String, Object> opts = new HashMap<>();
```

Comment:
Instead of the recursive call:
```
Arrays.asList(key.split("\\."))
      .stream()
      .parallel()
      .collect(Collector.of(HashMap<String, Object>::new,
        (left, right) -> left.put(right, value),
        (left, right) -> {
            left.put(left.keySet().iterator().next(), right);
            return left;
        }
      );
```

Reply:
Fixing the above:
```
    List<String> keys = Arrays.asList(key.split("\\."));
    Collections.reverse(keys);
    return keys.stream()
      .collect(Collector.of(HashMap<String, Object>::new,
        (left, right) -> {
          if (left.isEmpty()) {
            left.put(right, value);
          } else {
            HashMap<String, Object> l = new HashMap<>(left);
            left.clear();
            left.put(right, l);
          }
        },
        (left, right) -> left));
```
Doesn't look that hot anymore. Still interestiiiiing.